### PR TITLE
Add support for Azure DevOps pull request events

### LINKED ISSUE #80: Pull Request Trigger type support for Azure Devops
The Code Push webhook on Vsts (Azure Devops) work great with bitrise. But is there any way (a workaround) to be able to do a Pull Request rigger.
I mean I dont want to have every code push to any branch in our AzureDevops repo to trigger a build on bitrise. 

How to do that ?

### PR DESCRIPTION
**Pull request by DewaldDeJager**
### New Pull Request Checklist

- [X] Run `go fmt` on your files (e.g. `go fmt ./service/common.go`, or on the whole `service` folder: `go fmt ./service/...`)
- [X] Write tests for your code
  - The tests should cover both "success" and "error" cases.
  - The tests should also check **all the returned variables**, don't ignore any returned value!
  - Ideally the tests should be easily readable, we usually use tests to document our code instead of code comments.
    An example, if you'd write a comment like "Given X this function will return Y" or
    "Beware, if the input is X this function will return Y" then you should implement this as
    a unit test, instead of writing it as a comment.
- [X] If your Pull Request is more than a bug fix you should also check `README.md` and change/add the descriptions there - also
  feel free to add yourself as a contributor if you implement support for a new service ;)
- [X] Before creating the Pull Request you should also run `bitrise run test` with the [Bitrise CLI](https://www.bitrise.io/cli),
  to perform all the automatic checks (which will run on your Pull Request when you open it).

### Summary of Pull Request
Fixes #80 :) 

Typical dev flow (creating feature branch -> updating feature branch -> creating PR -> updating feature branch -> merging PR):
image

[Azure DevOps event reference](https://github.com/MicrosoftDocs/azure-devops-docs/blob/main/docs/service-hooks/events.md) (slightly out-of-date at time of writing)

This is my first time using Go so any feedback related to language/library usage would also be appreciated!

### CODE REVIEW

**Comment by DewaldDeJager** [comment]
@viktorbenei @sylank @gaborszakacs Please will you review this PR?

I can't access the PR status check logs. I've tested fairly thoroughly and am not getting any failures. I think it just failed because I converted it from a draft PR. If so, please will you rerun the checks? Otherwise, please let me know why they are failing.

**Comment by sylank** [comment]
@DewaldDeJager sorry for the delay, I checked the CI here is the error:
```
+------------------------------------------------------------------------------+
| (5) Go Lint                                                                  |
+------------------------------------------------------------------------------+
| id: script                                                                   |
| version: 1.2.0                                                               |
| collection: https://github.com/bitrise-io/bitrise-steplib.git                |
| toolkit: bash                                                                |
| time: 2022-05-25T13:29:49Z                                                   |
+------------------------------------------------------------------------------+
|                                                                              |
-> Linting: github.com/bitrise-io/bitrise-webhooks
-> Linting: github.com/bitrise-io/bitrise-webhooks/bitriseapi
-> Linting: github.com/bitrise-io/bitrise-webhooks/config
-> Linting: github.com/bitrise-io/bitrise-webhooks/metrics
-> Linting: github.com/bitrise-io/bitrise-webhooks/service
-> Linting: github.com/bitrise-io/bitrise-webhooks/service/hook
-> Linting: github.com/bitrise-io/bitrise-webhooks/service/hook/assembla
-> Linting: github.com/bitrise-io/bitrise-webhooks/service/hook/bitbucketserver
-> Linting: github.com/bitrise-io/bitrise-webhooks/service/hook/bitbucketv2
-> Linting: github.com/bitrise-io/bitrise-webhooks/service/hook/common
-> Linting: github.com/bitrise-io/bitrise-webhooks/service/hook/deveo
-> Linting: github.com/bitrise-io/bitrise-webhooks/service/hook/github
-> Linting: github.com/bitrise-io/bitrise-webhooks/service/hook/gitlab
-> Linting: github.com/bitrise-io/bitrise-webhooks/service/hook/gogs
-> Linting: github.com/bitrise-io/bitrise-webhooks/service/hook/passthrough
-> Linting: github.com/bitrise-io/bitrise-webhooks/service/hook/slack
-> Linting: github.com/bitrise-io/bitrise-webhooks/service/hook/visualstudioteamservices
=> Golint issues found:
/bitrise/go/src/github.com/bitrise-io/bitrise-webhooks/service/hook/visualstudioteamservices/visualstudioteamservices.go:22:2: exported const Push should have comment (or a comment on this block) or be unexported
/bitrise/go/src/github.com/bitrise-io/bitrise-webhooks/service/hook/visualstudioteamservices/visualstudioteamservices.go:62:2: struct field PullRequestId should be PullRequestID
|                                                                              |
+---+---------------------------------------------------------------+----------+
| x | Go Lint (exit code: 1)                                        | 0.93 sec |
+---+---------------------------------------------------------------+----------+
| Issue tracker: https://github.com/bitrise-io/steps-script/issues             |
| Source: https://github.com/bitrise-io/steps-script                           |
+---+---------------------------------------------------------------+----------+
```

**Comment by DewaldDeJager** [comment]
@sylank  Thanks! Let me tend to that right away.

Yes, I have done thorough integration testing with a self-hosted webhook server pointing to the prod Bitrise build API. The screenshot in the initial post is a real integration test between Azure DevOps and prod Bitrise.

File: service/hook/visualstudioteamservices/visualstudioteamservices.go [hunk 1]
```diff
@@ -17,42 +18,82 @@ const (
...
+	DetailedMessage EventMessage      `json:"detailedMessage"`
+	Message         EventMessage      `json:"message"`
+}
+
+// PullRequestEventModel ...
+type PullRequestEventModel struct {
+	SubscriptionID  string                   `json:"subscriptionId"`
+	EventType       string                   `json:"eventType"`
+	PublisherID     string                   `json:"publisherId"`
+	Resource        PullRequestResourceModel `json:"resource"`
```
[the remark this answers is not in the document]
Comment (sylank) [reply service/hook/visualstudioteamservices/visualstudioteamservices.go:93]: I think it is a good practice to have these two separated structures for these separated events. 
- When MS refactors the API you need to update only the PullRequest processing flow and structure - decoupled from the Push model.
- The structure represents only one responsibility. 

So it is ok that way 😺 .

**Comment by DewaldDeJager** [comment]
When running `bitrise run test` the `Err check` and `Go Lint` steps fail for me:

> /tmp/bitrise345110792/step_src/._script_cont: line 1: errcheck: command not found
> /tmp/bitrise515290330/step_src/._script_cont: line 6: golint: command not found

Which is weird because they are installed in the `_install_test_tools` workflow.

**Comment by sylank** [comment]
I think we should review this at Bitrise. Now the CI is green so I will merge this PR.

> This is my first time using Go so any feedback related to language/library usage would also be appreciated!
This project has not in an excellent condition and we have other kinds of Go standards at Bitrise. This project needs a refactor I think. So unfortunately it is not the best learning material :P 
This step is a kinda new project: https://github.com/bitrise-steplib/bitrise-step-artifact-pull
We have an internal package for web services it is not open source :(

We are reading/using this book: https://www.oreilly.com/library/view/learning-go/9781492077206/
And this: https://lets-go.alexedwards.net/

**Review by sylank — APPROVED** [review]

**Comment by sylank** [comment]
@DewaldDeJager deployed

**Comment by DewaldDeJager** [comment]
Working as expected 😄 
image

**Comment by DewaldDeJager** [comment]
> I think we should review this at Bitrise. Now the CI is green so I will merge this PR.
> 
> > This is my first time using Go so any feedback related to language/library usage would also be appreciated!
> 
> This project is not in excellent condition and we have other kinds of Go standards at Bitrise. This project needs a refactor I think. So unfortunately it is not the best learning material :P This step is a kinda new project: https://github.com/bitrise-steplib/bitrise-step-artifact-pull We have an internal package for web services it is not open source :(
> 
> We are reading/using this book: https://www.oreilly.com/library/view/learning-go/9781492077206/ And this: https://lets-go.alexedwards.net/

Thanks for the resources, it's much appreciated! I'll reference the artifact pull step and see if I can contribute to improving this project.

**Comment by viktorbenei** [comment]
Thanks @DewaldDeJager for your contribution!! 🚀

**Pull request merged**